### PR TITLE
[Merged by Bors] - Consolidated CRD to top level cluster config

### PR DESCRIPTION
**Pull request by maltesander**
# Description

This is a consolidation PR as prerequisite for logging.

fixes https://github.com/stackabletech/hive-operator/issues/291

## Review Checklist

- [ ] Code contains useful comments
- [ ] CRD change approved (or not applicable)
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -150,8 +161,6 @@ pub struct MetastoreStorageConfig {
 #[serde(rename_all = "camelCase")]
 pub struct MetaStoreConfig {
     pub warehouse_dir: Option<String>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:163]:
```suggestion
    /// The location of default database for the Hive warehouse.
    /// Maps to the `hive.metastore.warehouse.dir` setting.
    pub warehouse_dir: Option<String>,
```

File: rust/crd/src/lib.rs [hunk 2]
```diff
@@ -283,20 +292,20 @@ impl Configuration for MetaStoreConfig {
 
     fn compute_cli(
         &self,
-        _resource: &Self::Configurable,
+        resource: &Self::Configurable,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:295]: Nit: please rename to `hive`
Same goes for `compute_env` and `compute_files`

File: rust/operator-binary/src/controller.rs [hunk 3]
```diff
@@ -174,15 +174,16 @@ pub async fn reconcile_hive(hive: Arc<HiveCluster>, ctx: Arc<Ctx>) -> Result<Act
...
-        )
-    } else {
-        None
-    };
+    let s3_connection_spec: Option<S3ConnectionSpec> =
+        if let Some(s3) = &hive.spec.cluster_config.s3 {
+            Some(
+                s3.resolve(client, hive.namespace().as_deref().unwrap())
+                    .await
+                    .context(ResolveS3ConnectionSnafu)?,
```
Comment (sbernauer) [inline rust/operator-binary/src/controller.rs:182]: New error variant (I suggest to put it at top)
```
    #[snafu(display("object defines no namespace"))]
    ObjectHasNoNamespace,
```

```suggestion
                s3.resolve(
                    client,
                    &hive.namespace().ok_or(Error::ObjectHasNoNamespace)?,
                )
                .await
                .context(ResolveS3ConnectionSnafu)?,
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
Only some nit-picking.

Do you mind creating follow-up PRs for the depended products? I can only think of Trino (maybe Spark has a HMS test as well)

File: rust/crd/src/lib.rs (hunk 2 above)
Comment (maltesander) [reply rust/crd/src/lib.rs]: done

File: rust/crd/src/lib.rs (hunk 1 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:165]: done

**Review by sbernauer — APPROVED** [review]

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**